documentation/ImplementingASemanticCheck.md

```diff
@@ -0,0 +1,816 @@
+3. the relevant symbol passed as the actual argument.
+
+The first and third are needed since they're required to call the utility
+functions.  The second is needed to determine whether to call them.
+
+## Finding the source location
+The source code location information that I'd need for the error message must
+come from the parse tree.  I looked in the file
+`include/flang/parser/parse-tree.h` and determined that a `struct Expr`
+contained source location information since it had the field `CharBlock
+source`.  Thus, if I visited a `parser::Expr` node, I could get the source
+location information for the associated expression.
+
+## Determining the `INTENT`
+I knew that I could find the `INTENT` of the dummy argument associated with the
+actual argument from the function called `dummyIntent()` in the class
+`evaluate::ActualArgument` in the file `include/flang/evaluate/call.h`.  So
+if I could find an `evaluate::ActualArgument` in an expression, I could
+  determine the `INTENT` of the associated dummy argument.  I knew that it was
+  valid to call `dummyIntent()` because the data on which `dummyIntent()`
+  depends is established during semantic processing for expressions, and the
+  semantic processing for expressions happens before semantic checking for DO
+  constructs.
+
+In my prior work on checking the INTENT of arguments for SUBROUTINE calls,
+the parse tree held a node for the call (a `parser::CallStmt`) that contained
+an `evaluate::ProcedureRef` node.
+```C++
+  struct CallStmt {
+    WRAPPER_CLASS_BOILERPLATE(CallStmt, Call);
+    mutable std::unique_ptr<evaluate::ProcedureRef,
+        common::Deleter<evaluate::ProcedureRef>>
+        typedCall;  // filled by semantics
+  };
+```
+The `evaluate::ProcedureRef` contains a list of `evaluate::ActualArgument`
+nodes.  I could then find the INTENT of a dummy argument from the
+`evaluate::ActualArgument` node.
+
+For a FUNCTION call, though, there is no similar way to get from a parse tree
+node to an `evaluate::ProcedureRef` node.  But I knew that there was an
+existing framework used in DO construct semantic checking that traversed an
+`evaluate::Expr` node collecting `semantics::Symbol` nodes.  I guessed that I'd
+be able to use a similar framework to traverse an `evaluate::Expr`  node to
+find all of the `evaluate::ActualArgument` nodes.  
+
+All of the declarations associated with both FUNCTION and SUBROUTINE calls are
+in `include/flang/evaluate/call.h`.  An `evaluate::FunctionRef` inherits from
+an `evaluate::ProcedureRef` which contains the list of
+`evaluate::ActualArgument` nodes.  But the relationship between an
+`evaluate::FunctionRef` node and its associated arguments is not relevant.  I
+only needed to find the `evaluate::ActualArgument` nodes in an expression.
+They hold all of the information I needed.
+
+So my plan was to start with the `parser::Expr` node and extract its
+associated `evaluate::Expr` field.  I would then traverse the
```

Comment:
Do you think a little bit explanation between `parser::Expr` and `evaluate::Expr` would be better?

Reply:
Good point.  I was certainly confused when I first learned that there were two different types named `Expr` in the compiler.  I'll add a short explanation.